ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,339 @@
+   */
+  @Override
+  public void appendFrom(ReadableByteChannel channel, long size)
+      throws IOException {
+    ensureActive();
+    if (endOffset.get() + size > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException(
+          "Channel cannot be written to segment [" + file.getAbsolutePath() + "] because" + " it exceeds the capacity ["
+              + capacityInBytes + "]");
+    } else {
+      long bytesWritten = 0;
+      while (bytesWritten < size) {
+        bytesWritten += fileChannel.transferFrom(channel, endOffset.get() + bytesWritten, size - bytesWritten);
+      }
+      endOffset.addAndGet(bytesWritten);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * The read is not started if it cannot be completed.
+   * @param buffer The buffer into which the read needs to write to
+   * @param position The position to start the read from
+   * @throws IllegalArgumentException if {@code position} < 0 or > {@link #getEndOffset()} or if {@code buffer} size is
+   * greater than the data available for read.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  @Override
+  public void readInto(ByteBuffer buffer, long position)
+      throws IOException {
+    if (position < 0 || position > getEndOffset()) {
+      throw new IllegalArgumentException(
+          "Provided position [" + position + "] is out of bounds for the segment [" + file.getAbsolutePath()
+              + "] with end offset [" + getEndOffset() + "]");
+    }
+    if (position + buffer.remaining() > getEndOffset()) {
+      metrics.overflowReadError.inc();
+      throw new IllegalArgumentException("Cannot read from segment [" + file.getAbsolutePath() + "] from position [" +
+          position + "] for size [" + buffer.remaining() + "] because it exceeds the end offset [" + endOffset + "]");
+    }
+    long bytesRead = 0;
+    int size = buffer.remaining();
+    while (bytesRead < size) {
+      bytesRead += fileChannel.read(buffer, position + bytesRead);
+    }
+  }
+
+  /**
+   * Writes {@code size} number of bytes from the channel {@code channel} into the segment at {@code offset}.
+   * <p/>
+   * There is no state checking done for this function and the assumption is that the caller knows what is being done.
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param channel The channel from which data needs to be written from.
+   * @param offset The offset in the segment at which tho start writing.
+   * @param size The amount of data in bytes to be written from the channel.
+   * @throws IllegalArgumentException if {@code offset} < 0 or if there is not enough space for {@code offset } +
+   * {@code size} data.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   *
+   */
+  void writeFrom(ReadableByteChannel channel, long offset, long size)
```

Comment:
Can this be allowed in an ACTIVE segment. I thought its allowed only on a SEALED one. Are these to accommodate zeroing out the bytes when the segment moves from SEALED to FREE state ?